sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala

```diff
@@ -39,7 +39,9 @@ class ConfigBehaviorSuite extends QueryTest with SharedSQLContext {
     def computeChiSquareTest(): Double = {
       val n = 10000
       // Trigger a sort
-      val data = spark.range(0, n, 1, 1).sort('id.desc)
+      // Range has range partitioning in its output now. To have a range shuffle, we
+      // need to run a repartition first.
+      val data = spark.range(0, n, 1, 1).repartition(10).sort('id.desc)
```

Comment:
sorry, I am just curious, why is `sort('id.desc)` not causing a shuffle? Shouldn't it be ordered by `'id.asc` without the sort?

Reply:
This test requires a range shuffle. Previously `range` has unknown output partitioning/ordering, so there is a range shuffle inserted before `sort`.

For now `range` has an ordered output, so planner doesn't insert the shuffle we need here.

Reply:
I'm also confused here, the range output ordering is `'id.asc`, which doesn't match `'id.desc` how can we avoid shuffle?

Reply:
Because it is just one partition now.

Reply:
hmm, isn't `spark.range(0, n, 1, 10)` almost same as `spark.range(0, n, 1, 1).repartition(10)`?

Reply:
This is a good point.

This is query plan and partition size for `spark.range(0, n, 1, 1).repartition(10).sort('id.desc)`, when we set `SQLConf.RANGE_EXCHANGE_SAMPLE_SIZE_PER_PARTITION` to 1:

```
== Physical Plan ==
*(2) Sort [id#15L DESC NULLS LAST], true, 0
+- Exchange rangepartitioning(id#15L DESC NULLS LAST, 4)
   +- Exchange RoundRobinPartitioning(10)
      +- *(1) Range (0, 10000, step=1, splits=1)

1666, 3766, 2003, 2565
```

`spark.range(0, n, 1, 10).sort('id.desc)`:

```
== Physical Plan ==
*(2) Sort [id#13L DESC NULLS LAST], true, 0
+- Exchange rangepartitioning(id#13L DESC NULLS LAST, 4)
   +- *(1) Range (0, 10000, step=1, splits=10)

(2835, 2469, 2362, 2334)
```

Because `repartition` shuffles data with `RoundRobinPartitioning`, I guess that it makes the worse sampling for range exchange. Without `repartition`, `Range`'s output is already range partitioning, so it can get sampling leading better range boundaries.

Reply:
i see, so the `100` and `300` in this test are coupled with the physical execution. I feel the right way to test this is, instead of hardcoding `100` and `300`, we should have `a` and `b`, and check if `b > 3 * a` or something.

Reply:
By `spark.range(0, n, 1, 10).sort('id.desc)`, there is no 3 times liner relation between `a` and `b`. As shown above, this is also evenly distribution, the chi-sq value is also under `100`.

Here we need a redistribution on data to make sampling difficult. Previously, a repartition is added automatically before `sort`. Now `range` has correct output partition info, so the repattition must be added manually.